cli/azd/internal/vsrpc/handler.go

```diff
@@ -183,7 +218,12 @@ func HandlerFunc3[T1 any, T2 any, T3 any, TRet any](f func(context.Context, T1,
 			return reply(ctx, nil, err)
 		}
 
-		ret, err := f(ctx, t1, t2, t3)
+		ret, err := func() (ret TRet, err error) {
+			defer capturePanic(&err)
+			ret, err = f(ctx, t1, t2, t3)
+			return
```

Comment:
ah!  It took me a few seconds to get how using `return` was not missing `TRet, err` there.. This is because we use named outputs :)

I feel like taking a pointer to the error is forcing the existence an error before calling the function. And what is an error but just an interface, right?.  Have you consider taking a `func() error` instead?  like:

```go
// A case where `f` only returns an error:
err = capturePanic(func() error {
			return f(ctx, t1, t2, t3)
		})

// Case where `f` returns something and error
var ret TRet
err = capturePanic(func() error {
	innerRet, err := f(ctx, t1, t2, t3)
	ret = innerRet
	return err
})

// capture holds all the logic to wrap any function
func capturePanic(f func() error) error {
    var err error
    defer func() {
        if p := recover(); p != nil {
	        err = &jsonrpc2.Error{
		        Code:    jsonrpc2.InternalError,
		        Message: fmt.Sprintf("panic: %v", p),
	        }
        }
    }()
    return f()
}
```

Reply:
> ah! It took me a few seconds to get how using return was not missing TRet, err there.. This is because we use named outputs :)

Yes - this is sort of clever, but it felt like the most direct way to deal with the problem that was introduced by this non local control flow.

I tried moving to this model:

```go
// capture holds all the logic to wrap any function
func capturePanic(f func() error) error {
    var err error
    defer func() {
        if p := recover(); p != nil {
	        err = &jsonrpc2.Error{
		        Code:    jsonrpc2.InternalError,
		        Message: fmt.Sprintf("panic: %v", p),
	        }
        }
    }()
    return f()
}
```

But it does not work out of the box.  I had to make some small changes to the structure of this helper (the `return f()` is problematic - if `f()` panic'd then the thing getting returned was `nil` when we recovered, not the `error` assigned in our recovery handler.  I was able to fix that, but it felt like when I was done things were not "better" just "different".

> I feel like taking a pointer to the error is forcing the existence an error before calling the function. And what is an error but just an interface, right?.

Yeah - passing a pointer to `error` does seem a little strange here - but I do feel this is pretty idomatic - `capturePanic` arranges things such that if a panic happens the error you pass a pointer to is updated so the value is a representation of the panic.

I personally feel like this shape:

```go
// Case where `f` returns something and error
var ret TRet
err = capturePanic(func() error {
	innerRet, err := f(ctx, t1, t2, t3)
	ret = innerRet
	return err
})
```

Is harder to grok than this shape:

```go
ret, err := func() (ret TRet, err error) {
	defer capturePanic(&err)
	ret, err = f(ctx, t1)
	return
}()
```

To me it's clearer to see the return value and the error get assigned in this way and my eyes can sort of "gloss over" the body of the anonymous function that we are invoking.  I could have worked around this by adding a `capturePanicAndValue` method that returned both values - but then we have two helpers.

I did take some time to explore your pattern - but I think I prefer what we have here right now a bit more.

I do feel like the exercise was useful to me - so that's for taking the time to raise the issue :-).